File: slidingWindow/findMedianSortedArrays.js
/*
There are two sorted arrays nums1 and nums2 of size m and n respectively.

Find the median of the two sorted arrays. The overall run time complexity should be O(log(m + n))

You may assume nums1 and nums2 cannot be both empty.

Example 1:

  nums1 = [1, 3]
  nums2 = [2]

  The median is 2.0

Example 2:

  nums1 = [1,2]
  nums2 = [3,4]

  The median is (2 + 3) / 2 = 2.5
*/

const findMedianSortedArrays = (nums1, nums2) => {
  // perform on shorter of two arrays
  if (nums1.length > nums2.length) return findMedianSortedArrays(nums2, nums1);

  const x = nums1.length;
  const y = nums2.length;

  let lo = 0;
  let hi = x;

  while (lo <= hi) {
    let partitionX = ((lo + hi) / 2) | 0;
    let partitionY = ((x + y + 1) / 2 - partitionX) | 0;

    let maxLeftX = partitionX === 0 ? -Infinity : nums1[partitionX - 1];
    let minRightX = partitionX === x ? Infinity : nums1[partitionX];

    let maxLeftY = partitionY === 0 ? -Infinity : nums2[partitionY - 1];
    let minRightY = partitionY === y ? Infinity : nums2[partitionY];

    if (maxLeftX <= minRightY && maxLeftY <= minRightX) {
      if ((x + y) & 1) return Math.max(maxLeftX, maxLeftY); // odd length
      return (
        (Math.max(maxLeftX, maxLeftY) + Math.min(minRightX, minRightY)) / 2
      ); // even length

      // binary search - move closer left
    } else if (maxLeftX > minRightY) hi = partitionX - 1;
    else lo = partitionX + 1; // move closer right
  }
};

File: slidingWindow/findMedianSortedArrays.py
'''
There are two sorted arrays nums1 and nums2 of size m and n respectively.

Find the median of the two sorted arrays. The overall run time complexity should be O(log(m + n))

You may assume nums1 and nums2 cannot be both empty.

Example 1:

  nums1 = [1, 3]
  nums2 = [2]

  The median is 2.0

Example 2:

  nums1 = [1,2]
  nums2 = [3,4]

  The median is (2 + 3) / 2 = 2.5
'''


def findMedianSortedArrays(nums1, nums2):
    if len(nums1) > len(nums2):
        return findMedianSortedArrays(nums2, nums1)

    x, y = len(nums1), len(nums2)

    lo, hi = 0, x

    while lo <= hi:
        partX = (lo + hi) // 2  # partition for shorter set
        # partition for longer set including partX
        partY = ((x + y + 1)) // 2 - partX

        # create boundaries
        maxLeftX = float("-inf") if partX == 0 else nums1[partX - 1]
        minRightX = float("inf") if partX == x else nums1[partX]

        maxLeftY = float("-inf") if partY == 0 else nums2[partY - 1]
        minRightY = float("inf") if partY == y else nums2[partY]

        if maxLeftX <= minRightY and maxLeftY <= minRightX:
            if (x + y) % 2 == 0:
                # even length
                return float(max(maxLeftX, maxLeftY) + min(minRightX, minRightY)) / 2
            else:
                # odd length
                return float(max(maxLeftX, maxLeftY))
        elif maxLeftX > minRightY:
            hi = partX - 1
        else:
            lo = partX + 1
